File: problem-solving/_lc-278-easy-first-bad-version/1-linear-search.py
"""
https://leetcode.com/problems/first-bad-version/

Complexity
----------
Time: O(n)
Space: O(1)
"""

# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        for v in range(1, n + 1):
            if isBadVersion(v):
                return v

File: problem-solving/_lc-74-medium-search-a-2d-matrix/1-linear-search.go
/*
https://leetcode.com/problems/search-a-2d-matrix/

Complexity
----------
Time: O(rows + cols)
Space: O(1)
*/

func searchMatrix(matrix [][]int, target int) bool {
    r := 0
    c := len(matrix[0]) - 1
    for r < len(matrix) && c >= 0 {
        if matrix[r][c] == target {
            return true
        } else if matrix[r][c] < target {
            r += 1
        } else {
            c -= 1
        }
    }
    return false
}
